File: Python/maxSubArray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        Time Complexity:O(n) Space Complexity O(1)
        Kadane's Algorithm
        DP method: Keeping compare between maxPre plus currenty number and current number,
        get the larger one, and then campare to recorded max value,
        update that max value, and finally update maxPre.
        """
##        if not nums: return 0
        maxHere = maxSoFar = nums[0]
        for i in nums[1:]:
            maxHere = max(i, maxHere + i)
            maxSoFar = max(maxHere, maxSoFar)
        return maxSoFar
        

File: src/leetcode/S0053_Maximum Subarray/maxSubArray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  if (!nums.length) return 0;

  // max 是全局的最大值
  let max = nums[0];
  // sum 是遍历到当前的加和,累加
  let sum = nums[0];

  for (let i = 1; i < nums.length; i++) {
    // 如果 sum 加当前数更小了，说明不应该加，直接用当前数重置
    // 如果 sum 加当前数更大了，说明找对了，继续向后
    sum = Math.max(nums[i], sum + nums[i]);
    max = Math.max(sum, max);
  }
  return max;
};
